<head>
<!-- Generated by perlmod2www.pl -->
<title>
PROTEO documentation.
</title>
<link rel="stylesheet" type="text/css" href="../perl.css">
</head>
<body bgcolor="white">
<span id="modRootPath">APPRIS::Export</span>
<span id="modName">PROTEO</span>
<div id="sep"></div>
<a name="TOP" id="TOP"></a>
<table id="daArea" width="100%"><tr><td id="daEntry" ><a href="#SUMMARY">Summary</a></td><td id="daEntry" ><a href="#libs">Included libraries</a></td><td id="daEntry" ><a href="#vardefs">Package variables</a></td><td id="daEntry" ><a href="#SYNOPSIS">Synopsis</a></td><td id="daEntry" ><a href="#DESCRIPTION">Description</a></td><td id="daEntry" ><a href="#General">General documentation</a></td><td id="daEntry" ><a href="#Methods">Methods</a></td></tr></table>
<div id="sep"></div>
<div id="descTitle">Summary</div>
<div id="descArea"><B>APPRIS::Export::PROTEO</B> - Utility functions for info exporting</div>
<div id="sep"></div>
<a name="vardefs" id="vardefs"></a><div id="descTitle">Package variables</div>
<div id="descArea"><div id="libEntry">No package variables defined.</div>
</div>
<div id="sep"></div>
<div id="descTitle">Included modules</div>
<div id="descArea"><div id="libEntry"><span id="use"><a href="../Utils/Constant.html">APPRIS::Utils::Constant</a></span> qw ( $OK_LABEL $NO_LABEL $UNKNOWN_LABEL ) </div>
<div id="libEntry"><span id="use"><a href="../Utils/Exception.html">APPRIS::Utils::Exception</a></span> qw ( throw warning deprecate ) </div>
<div id="libEntry"><span id="use"><a href="../Utils/File.html">APPRIS::Utils::File</a></span> qw ( getTotalStringFromFile ) </div>
<div id="libEntry"><span id="use">Bio::Seq</span> </div>
<div id="libEntry"><span id="use">Bio::SeqIO</span> </div>
<div id="libEntry"><span id="use">Data::Dumper</span> </div>
<div id="libEntry"><span id="use"><a href="../Exporter.html">Exporter</a></span></div>
<div id="libEntry"><span id="use">FindBin</span></div>
</div>
<div id="sep"></div>
<a name="libs" id="libs"></a>
<div id="descTitle">Inherit</div>
<div id="descArea"><span id="isa"><a href="../Exporter.html">Exporter</a></span> </div>
<div id="sep"></div>
<a name="SYNOPSIS" id="SYNOPSIS"></a>
<div id="descTitle">Synopsis</div>
<div id="descArea"><pre id="podParagraph">  use APPRIS::Export::PROTEO
    qw(
       get_trans_annotations
     );

  or to get all methods just

  use APPRIS::Export::PROTEO;

  eval { get_trans_annotations($feature,$version) };
  if ($@) {
    print "Caught exception:\n$@";
  }</pre>
</div>
<div id="sep"></div>
<a name="DESCRIPTION" id="DESCRIPTION"></a>
<div id="descTitle">Description</div>
<div id="descArea">Retrieves information of transcript as GTF format.
</div>
<div id="sep"></div>
<a name="Methods" id="Methods"></a>
<div id="descTitle">Methods</div>
<table id="methIndArea" width="100%" cellspacing="0"><tr><td id="methIndEntry">get_appris_annotations2</td><td id="methIndEntry"><div id="error">No description</div></td><td id="methIndEntry"><a href="#CODE1">Code</a></td></tr>
<tr><td id="methIndEntry">get_proteo_annotations</td><td id="methIndEntry"><a href="#POD1">Description</a></td><td id="methIndEntry"><a href="#CODE2">Code</a></td></tr>
<tr><td id="methIndEntry">get_proteo_report($)</td><td id="methIndEntry"><div id="error">No description</div></td><td id="methIndEntry"><a href="#CODE3">Code</a></td></tr>
<tr><td id="methIndEntry">print_annotations</td><td id="methIndEntry"><a href="#POD2">Description</a></td><td id="methIndEntry"><a href="#CODE4">Code</a></td></tr>
</table>
<div id="sep"></div>
<a name="MethDesc" id="MethDesc"></a>
<div id="mainTitle">Methods description</div>
<div id="sep"></div>
<table id="methDescTitle" width="100%"><tr><td WIDTH="200"><a name="_pod_get_proteo_annotations" id="_pod_get_proteo_annotations"></a><a name="POD1" id="POD1"></a>get_proteo_annotations</td><td ><a href="#CODE2">code</a></td><td >&nbsp;&nbsp;&nbsp;&nbsp;</td><td ><a href="#POD2">next</a></td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<table id="methDescArea" width="100%"><tr><td><pre id="podParagraph">  Arg [1]    : String - the stable identifier of transcript
  Arg [2]    : String - the stable identifier of gene
  Arg [3]    : String - the external database name associated with transcript
  Arg [4]    : APPRIS::Transcript
  Arg [5]    : String - $version
  Example    : $annot = get_proteo_annotations($trans_id, $gen_id, $ext_id, $feat, $v);
  Description: Retrieves specific annotation.
  Returntype : String or undef</pre></td></tr></table>
<div id="sep"></div>
<table id="methDescTitle" width="100%"><tr><td WIDTH="200"><a name="_pod_print_annotations" id="_pod_print_annotations"></a><a name="POD2" id="POD2"></a>print_annotations</td><td ><a href="#CODE4">code</a></td><td ><a href="#POD1">prev</a></td><td ><a href="#POD3">next</a></td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<table id="methDescArea" width="100%"><tr><td><pre id="podParagraph">  Arg [1]    : String - common attributes
  Arg [2]    : String - optional attributes
  Example    : $annot = print_annotations($feature,$version);
  Description: Print the (common and optional) annotations.
  Returntype : String or undef</pre></td></tr></table>
<div id="sep"></div>
<a name="MethCode" id="MethCode"></a>
<div id="mainTitle">Methods code</div>
<div id="sep"></div>
<table id="methCodeTitle" width="100%"><tr><td width="200"><a name="CODE1" id="CODE1"></a>get_appris_annotations2</td><td >description</td><td >prev</td><td ><a href="#CODE2">next</a></td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="methCodeArea"><td ><span id="sub">sub</span> <span id="subName">get_appris_annotations2</span>
{<pre>	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$transcript_id</span>, <span id="var">$gene_id</span>, <span id="var">$external_id</span>, <span id="var">$feature</span>, <span id="var">$method_type</span>, <span id="var">$method_score</span>, <span id="var">$appris_annot</span>, <span id="var">$version</span><span id="braces">)</span> = <span id="var">@_</span>;
	
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$output</span><span id="braces">)</span> = <span id="sqStr">''</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_phase</span><span id="braces">)</span> = <span id="sqStr">'.'</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_source</span><span id="braces">)</span> = <span id="var">$GTF_CONSTANTS</span>-&gt;<span id="braces">{</span><span id="sqStr">'appris'</span><span id="braces">}</span>-&gt;<span id="braces">{</span><span id="sqStr">'source'</span><span id="braces">}</span>;

	<span id="comment"># Common attributes</span><br />	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$common</span><span id="braces">)</span> = <span id="braces">{</span>
			<span id="sqStr">'seqname'</span>	=&gt; <span id="var">$feature</span>-&gt;<span id="method">chromosome</span>,
			<span id="sqStr">'source'</span>	=&gt; <span id="var">$method_source</span>,
			<span id="sqStr">'type'</span>		=&gt; <span id="var">$method_type</span>,
			<span id="sqStr">'start'</span>		=&gt; <span id="var">$feature</span>-&gt;<span id="method">start</span>,
			<span id="sqStr">'end'</span>		=&gt; <span id="var">$feature</span>-&gt;<span id="method">end</span>,
			<span id="sqStr">'score'</span>		=&gt; <span id="var">$method_score</span>,
			<span id="sqStr">'strand'</span>	=&gt; <span id="var">$feature</span>-&gt;<span id="method">strand</span>,
			<span id="sqStr">'phase'</span>		=&gt; <span id="var">$method_phase</span>
	<span id="braces">}</span>;
	
	<span id="comment"># Optinal attributes</span><br />	<span id="perlKey">my</span><span id="braces">(</span><span id="var">$optional</span><span id="braces">)</span>;
	<span id="comment">#$optional-&gt;{'annotation'}		= $appris_annot;</span><br />	<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'gene_id'</span><span id="braces">}</span>			= <span id="var">$gene_id</span>;
	<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'transcript_id'</span><span id="braces">}</span>	= <span id="var">$transcript_id</span>;
	<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'transcript_name'</span><span id="braces">}</span>	= <span id="var">$external_id</span>;
	<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'version'</span><span id="braces">}</span>			= <span id="var">$version</span>;		
	<span id="perlKey">if</span> <span id="braces">(</span><span id="perlFunc">defined</span> <span id="var">$common</span> and <span id="perlFunc">defined</span> <span id="var">$optional</span><span id="braces">)</span> <span id="braces">{</span>
		<span id="var">$output</span> .= <span id="method">print_annotations</span><span id="braces">(</span><span id="var">$common</span>,<span id="var">$optional</span><span id="braces">)</span>;			
	<span id="braces">}</span>
	
	<span id="perlKey">return</span> <span id="var">$output</span>;
	
<span id="braces">}</span> <span id="comment"># End get_appris_annotations2</span><br /></pre>}</td></div>
<div id="sep"></div>
<table id="methCodeTitle" width="100%"><tr><td width="200"><a name="CODE2" id="CODE2"></a>get_proteo_annotations</td><td ><a href="#POD1">description</a></td><td ><a href="#CODE1">prev</a></td><td ><a href="#CODE3">next</a></td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="methCodeArea"><td ><span id="sub">sub</span> <span id="subName">get_proteo_annotations</span>
{<pre>	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$transcript_id</span>, <span id="var">$gene_id</span>, <span id="var">$external_id</span>, <span id="var">$feature</span>, <span id="var">$version</span><span id="braces">)</span> = <span id="var">@_</span>;

    <span id="perlKey">my</span> <span id="braces">(</span><span id="var">$output</span><span id="braces">)</span> = <span id="sqStr">''</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_score</span><span id="braces">)</span> = 0;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_phase</span><span id="braces">)</span> = <span id="sqStr">'.'</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_source</span><span id="braces">)</span> = <span id="var">$GTF_CONSTANTS</span>-&gt;<span id="braces">{</span><span id="sqStr">'proteo'</span><span id="braces">}</span>-&gt;<span id="braces">{</span><span id="sqStr">'source'</span><span id="braces">}</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$method_type</span><span id="braces">)</span> = <span id="var">$GTF_CONSTANTS</span>-&gt;<span id="braces">{</span><span id="sqStr">'proteo'</span><span id="braces">}</span>-&gt;<span id="braces">{</span><span id="sqStr">'type'</span><span id="braces">}</span>;

	<span id="comment"># Get annotations from file</span><br />	<span id="perlKey">if</span> <span id="braces">(</span> -e <span id="var">$RST_FILE</span> and <span id="braces">(</span>-s <span id="var">$RST_FILE</span> &gt; 0<span id="braces">)</span> <span id="braces">)</span> <span id="braces">{</span>
		
		<span id="comment"># get cont report</span><br />		<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_report</span><span id="braces">)</span> = <span id="method">get_proteo_report</span><span id="braces">(</span><span id="var">$transcript_id</span><span id="braces">)</span>;
		
		<span id="perlKey">if</span> <span id="braces">(</span> <span id="perlFunc">defined</span> <span id="var">$rst_report</span> and <span id="perlFunc">exists</span> <span id="var">$rst_report</span>-&gt;<span id="braces">{</span><span id="var">$transcript_id</span><span id="braces">}</span> <span id="braces">)</span> <span id="braces">{</span>
							
			<span id="var">$method_score</span> = <span id="var">$rst_report</span>-&gt;<span id="braces">{</span><span id="var">$transcript_id</span><span id="braces">}</span>-&gt;<span id="braces">{</span><span id="sqStr">'uniq_pep_num'</span><span id="braces">}</span>;
			
			<span id="comment"># common attributes</span><br />			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$common</span><span id="braces">)</span> = <span id="braces">{</span>
					<span id="sqStr">'seqname'</span>	=&gt; <span id="var">$feature</span>-&gt;<span id="method">chromosome</span>,
					<span id="sqStr">'source'</span>	=&gt; <span id="var">$method_source</span>,
					<span id="sqStr">'type'</span>		=&gt; <span id="var">$method_type</span>,
					<span id="sqStr">'start'</span>		=&gt; <span id="var">$feature</span>-&gt;<span id="method">start</span>,
					<span id="sqStr">'end'</span>		=&gt; <span id="var">$feature</span>-&gt;<span id="method">end</span>,
					<span id="sqStr">'score'</span>		=&gt; <span id="var">$method_score</span>,
					<span id="sqStr">'strand'</span>	=&gt; <span id="var">$feature</span>-&gt;<span id="method">strand</span>,
					<span id="sqStr">'phase'</span>		=&gt; <span id="var">$method_phase</span>
			<span id="braces">}</span>;
			<span id="comment"># optinal attributes</span><br />			<span id="perlKey">my</span><span id="braces">(</span><span id="var">$optional</span><span id="braces">)</span>;
			<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'gene_id'</span><span id="braces">}</span>			= <span id="var">$gene_id</span>;
			<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'transcript_id'</span><span id="braces">}</span>	= <span id="var">$transcript_id</span>;
			<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'transcript_name'</span><span id="braces">}</span>	= <span id="var">$external_id</span>;
			<span id="var">$optional</span>-&gt;<span id="braces">{</span><span id="sqStr">'version'</span><span id="braces">}</span>			= <span id="var">$version</span>;
			<span id="perlKey">if</span> <span id="braces">(</span><span id="perlFunc">defined</span> <span id="var">$common</span> and <span id="perlFunc">defined</span> <span id="var">$optional</span><span id="braces">)</span> <span id="braces">{</span>
				<span id="var">$output</span> .= <span id="method">print_annotations</span><span id="braces">(</span><span id="var">$common</span>,<span id="var">$optional</span><span id="braces">)</span>;			
			<span id="braces">}</span>				
			<span id="comment"># get appris annotation</span><br />			<span id="var">$output</span> .= <span id="method">get_appris_annotations2</span><span id="braces">(</span>	<span id="var">$transcript_id</span>,
												<span id="var">$gene_id</span>,
												<span id="var">$external_id</span>,
												<span id="var">$feature</span>,
												<span id="var">$method_type</span>,
												<span id="var">$method_score</span>,
												<span id="sqStr">''</span>,
												<span id="var">$version</span><span id="braces">)</span>;				
		<span id="braces">}</span>
	<span id="braces">}</span>
	
	<span id="perlKey">return</span> <span id="var">$output</span>;
	
<span id="braces">}</span> <span id="comment"># End get_proteo_annotations</span><br /></pre>}</td></div>
<div id="sep"></div>
<table id="methCodeTitle" width="100%"><tr><td width="200"><a name="CODE3" id="CODE3"></a>get_proteo_report($)</td><td >description</td><td ><a href="#CODE2">prev</a></td><td ><a href="#CODE4">next</a></td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="methCodeArea"><td ><span id="sub">sub</span> <span id="subName">get_proteo_report($)</span>
{<pre>	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$transcript_id</span><span id="braces">)</span> = <span id="var">@_</span>;
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_report</span><span id="braces">)</span>;		
	
	<span id="comment"># get cont report</span><br />	<span id="comment">#my ($rst_content) = APPRIS::Utils::File::getTotalStringFromFile($RST_FILE);</span><br />	<span id="perlKey">my</span><span id="braces">(</span><span id="var">@rst_content</span><span id="braces">)</span>;
	<span id="perlKey">eval</span> <span id="braces">{</span>
		<span id="var">@rst_content</span>=`<span id="perlFunc">grep</span> <span id="dqStr">"^<span id="var">$transcript_id</span>"</span> <span id="var">$RST_FILE</span>`;
	<span id="braces">}</span>;
	<span id="perlKey">return</span> <span id="var">$rst_report</span> <span id="perlKey">if</span><span id="braces">(</span>$@<span id="braces">)</span>;		
	
	<span id="comment">#foreach my $rst_cont ( @{$rst_content} ) {</span><br />	<span id="perlKey">foreach</span> <span id="perlKey">my</span> <span id="var">$rst_cont</span> <span id="braces">(</span> <span id="var">@rst_content</span> <span id="braces">)</span> <span id="braces">{</span>			
		<span id="perlKey">if</span> <span id="braces">(</span> <span id="var">$rst_cont</span> =~ <span id="regExp">/([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\t]*)\t+([^\$]*)$/</span> <span id="braces">)</span> <span id="braces">{</span>
			
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_trans_id</span><span id="braces">)</span> = <span id="var">$1</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_gene_id</span><span id="braces">)</span> = <span id="var">$2</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_num_trans</span><span id="braces">)</span> = <span id="var">$3</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_appris_pred</span><span id="braces">)</span> = <span id="var">$4</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_trans_length</span><span id="braces">)</span> = <span id="var">$5</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_uniq_pep_num</span><span id="braces">)</span> = <span id="var">$6</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_uniq_coverage</span><span id="braces">)</span> = <span id="var">$7</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_shared_pep_num</span><span id="braces">)</span> = <span id="var">$8</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_shared_coverage</span><span id="braces">)</span> = <span id="var">$9</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_uniq_pep_names</span><span id="braces">)</span> = <span id="var">$10</span>;
			<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$rst_shared_pep_names</span><span id="braces">)</span> = <span id="var">$11</span>;
			
			<span id="var">$rst_report</span>-&gt;<span id="braces">{</span><span id="var">$rst_trans_id</span><span id="braces">}</span> = <span id="braces">{</span>
				<span id="sqStr">'gene_id'</span>			=&gt; <span id="var">$rst_gene_id</span>,
				<span id="sqStr">'uniq_pep_num'</span>		=&gt; <span id="var">$rst_uniq_pep_num</span>,
				<span id="sqStr">'uniq_coverage'</span>		=&gt; <span id="var">$rst_uniq_coverage</span>,
				<span id="sqStr">'shared_pep_num'</span>	=&gt; <span id="var">$rst_shared_pep_num</span>,
				<span id="sqStr">'shared_coverage'</span>	=&gt; <span id="var">$rst_shared_coverage</span>,
				<span id="sqStr">'uniq_pep_names'</span>	=&gt; <span id="var">$rst_uniq_pep_names</span>,
				<span id="sqStr">'shared_pep_names'</span>	=&gt; <span id="var">$rst_shared_pep_names</span>,
			<span id="braces">}</span>
		<span id="braces">}</span>			
	<span id="braces">}</span>
	<span id="perlKey">return</span> <span id="var">$rst_report</span>;
	
<span id="braces">}</span> <span id="comment"># End get_proteo_report</span><br /></pre>}</td></div>
<div id="sep"></div>
<table id="methCodeTitle" width="100%"><tr><td width="200"><a name="CODE4" id="CODE4"></a>print_annotations</td><td ><a href="#POD2">description</a></td><td ><a href="#CODE3">prev</a></td><td >next</td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="methCodeArea"><td ><span id="sub">sub</span> <span id="subName">print_annotations</span>
{<pre>	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$common</span>, <span id="var">$optional</span><span id="braces">)</span> = <span id="var">@_</span>;
	
	<span id="perlKey">my</span> <span id="braces">(</span><span id="var">$output</span><span id="braces">)</span>=<span id="sqStr">''</span>;

	<span id="comment"># &lt;seqname&gt; &lt;source&gt; &lt;feature&gt; &lt;start&gt; &lt;end&gt;  &lt;score&gt; &lt;strand&gt; &lt;frame&gt; [attributes] [comments]</span><br />	<span id="comment">#  Print feature: Common feature</span><br />	<span id="var">$output</span> .= 	<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'seqname'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'source'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'type'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'start'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'end'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'score'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'strand'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>.
				<span id="var">$common</span>-&gt;<span id="braces">{</span><span id="sqStr">'phase'</span><span id="braces">}</span>.<span id="dqStr">"\t"</span>;

	<span id="comment">#  Print feature</span><br />	<span id="perlKey">while</span> <span id="braces">(</span> <span id="perlKey">my</span> <span id="braces">(</span><span id="var">$key</span>,<span id="var">$value</span><span id="braces">)</span> = <span id="perlFunc">each</span><span id="braces">(</span>%<span id="braces">{</span><span id="var">$optional</span><span id="braces">}</span><span id="braces">)</span> <span id="braces">)</span>
	<span id="braces">{</span>
		<span id="var">$output</span> .= <span id="var">$key</span>.<span id="sqStr">' &quot;'</span>.<span id="var">$value</span>.<span id="sqStr">'&quot;; '</span>;
	<span id="braces">}</span>
	<span id="var">$output</span>=~s<span id="regExp">/\; $/</span>\n/;<br />	
	<span id="perlKey">return</span> <span id="var">$output</span>;</pre>}</td></div>
<div id="sep"></div>
<a name="General" id="General"></a>
<div id="mainTitle">General documentation</div>
<div id="sep"></div>
<table id="genInfoTitle" width="100%"><tr><td><a name="_pod_CONTACT" id="_pod_CONTACT"></a>CONTACT</td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="genInfoArea"><pre id="podParagraph">  Please email comments or questions to the public INB
  developers list at <<a href="mailto:inb-tecnico@lists.cnio.es" id="podEmail">inb-tecnico@lists.cnio.es</a>&gt;.<br />
  Questions may also be sent to the developer, 
  Jose Manuel Rodriguez <<a href="mailto:jmrodriguez@cnio.es" id="podEmail">jmrodriguez@cnio.es</a>&gt;.</pre></div>
<div id="sep"></div>
</body>
</html>
